src/module_wrap.cc

```diff
@@ -115,6 +115,215 @@ ModuleCacheKey ModuleCacheKey::From(Local<Context> context,
       context, v8_request->GetSpecifier(), v8_request->GetImportAttributes());
 }
 
+// static
+thread_local ModuleInstantiationContext*
+    ModuleInstantiationContext::thread_local_context_;
+
+// static
+MaybeLocal<Module> ModuleInstantiationContext::ResolveModuleCallback(
+    Local<Context> context,
+    Local<String> specifier,
+    Local<FixedArray> import_attributes,
+    Local<Module> referrer) {
+  CHECK_NOT_NULL(thread_local_context_);
+  return thread_local_context_->ResolveModule(
+      context, specifier, import_attributes, referrer);
+}
+
+// static
+MaybeLocal<Object> ModuleInstantiationContext::ResolveSourceCallback(
+    Local<Context> context,
+    Local<String> specifier,
+    Local<FixedArray> import_attributes,
+    Local<Module> referrer) {
+  CHECK_NOT_NULL(thread_local_context_);
+  return thread_local_context_->ResolveSource(
+      context, specifier, import_attributes, referrer);
+}
+
+ModuleInstantiationContext::ModuleInstantiationContext() {
+  // Only one ModuleInstantiationContext can exist per thread at a time.
+  CHECK_NULL(thread_local_context_);
+  thread_local_context_ = this;
+}
+
+ModuleInstantiationContext::~ModuleInstantiationContext() {
+  // Ensure that the thread-local context is this context.
+  CHECK_EQ(thread_local_context_, this);
+  thread_local_context_ = nullptr;
+}
+
+MaybeLocal<Module> ModuleInstantiationContext::ResolveModule(
+    Local<Context> context,
+    Local<String> specifier,
+    Local<FixedArray> import_attributes,
+    Local<Module> referrer) {
+  Isolate* isolate = context->GetIsolate();
+  Environment* env = Environment::GetCurrent(context);
+  if (env == nullptr) {
+    THROW_ERR_EXECUTION_ENVIRONMENT_NOT_AVAILABLE(isolate);
+    return MaybeLocal<Module>();
+  }
+
+  ModuleCacheKey cache_key =
+      ModuleCacheKey::From(context, specifier, import_attributes);
+
+  ModuleWrap* dependent = ModuleWrap::GetFromModule(env, referrer);
+  if (dependent == nullptr) {
+    THROW_ERR_VM_MODULE_LINK_FAILURE(
+        env, "request for '%s' is from invalid module", cache_key.specifier);
+    return MaybeLocal<Module>();
+  }
+  if (!dependent->IsLinked()) {
+    THROW_ERR_VM_MODULE_LINK_FAILURE(
+        env,
+        "request for '%s' is from a module not been linked",
+        cache_key.specifier);
+    return MaybeLocal<Module>();
+  }
+
+  ResolveCache* resolve_cache;
+  if (!GetModuleResolveCache(env, dependent, &resolve_cache).IsJust()) {
```

Comment:
Cross referencing https://github.com/nodejs/node/pull/59000#discussion_r2219184583 - I think here we don't need the map with the globals, we can just have a string -> size_t map where the keys are serialized module requests, and the values are indexes to the modules array. Then  `ResolveModule` can just look up the index again, and then index into the kLinkedRequestsSlot. That way we don't need to create globals - no matter on the stack or on the heap, since our look up target is "index in the modules array", not "the modules".